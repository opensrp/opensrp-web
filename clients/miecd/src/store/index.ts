import { connectReducer, getConnectedStore } from '@onaio/connected-reducer-registry';
import { gateKeeperReducer, gateKeeperReducerName } from '@onaio/gatekeeper';
import reducerRegistry, { combine, Registry } from '@onaio/redux-reducer-registry';
import session, { reducerName as sessionReducer } from '@onaio/session-reducer';

/** Initial reducers in the reducer registry */
const defaultReducers: Registry = {
    router: connectReducer as any /** Dirty hack because connectRouter LocationChangeAction does not extend Redux.AnyAction */,
};

/** Add users reducer to registry */
defaultReducers[sessionReducer] = session;
defaultReducers[gateKeeperReducerName] = gateKeeperReducer;

/** The initial store for the reveal web app */
const store = getConnectedStore(defaultReducers);

/** Set listener to add reducers to store when registered */
reducerRegistry.setChangeListener(reducers => {
    store.replaceReducer(combine(reducers));
});

export default store;
