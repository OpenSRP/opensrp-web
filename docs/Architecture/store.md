# overview

This documentation will take you through the process of creating and working with:

    - reducers
    - actions
    - selectors

## Where do I find the store

the store reducer modules are found within src/store. The initial directory structure looks something like this: (Note: this structure only depicts the initial directory structure and will as expected contain more directories and files as more are added with time)

```
...
├── src
│   ├── ...
│   ├── store
│   │   ├── ducks
│   │   │    ├── superset.ts
│   │   │    ├── tests/
│   │   ├── tests
│   │   ├── index.ts
│   │   └── selectors.ts
...
```

A single reducer module will contain the afore mentioned parts i.e. its respective actions, action creators if any, its selectors and the reducer function

## Actions

An action can be any object with a type property, they are the only way provisioned for changing the state in a redux store.

### Adding Actions

Its a common convention as well as recommended one to Use type constants, [read more on this here]().

steps:

- Add An interface describing the action structure: redux exposes an `AnyAction` type that can be extended to write custom action interfaces
- Create a general actiontypes to be used to type check all actions that will be passed to the reducer function as an argument. this will be in the form of [intersection type]()

## Selectors

whereas the action puts data in the store; the selectors on the other hand act as getters of the slice of data that this reducer module handles

selectors are functions that will take at least one argument, the store, Then based on any other arguments like say an id, can be used to filter down the data retrieved.

## The Reducer

The reducer function, takes 2 arguments; the state and an action.
using the type property of the action, the reducer should then know how to modify the state.
such functionality could be implemented as a form of switch strcture such as a list of if statements or the more recommended approach which is to use the actual switch statement

### Adding the Reducer

- Define an interface for the data to be stored in the store
- Define an interface for the store state.
- write the actual reducer function, while applying the required type annotations as required.
