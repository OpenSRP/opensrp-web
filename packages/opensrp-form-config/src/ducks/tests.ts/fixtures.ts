export const fixManifestReleases = [
    {
        identifier: '1.0.11',
        json: '{"forms_version":"1.0.11","identifiers":["blood_screening.json"}',
        appId: 'org.smartregister.reveal',
        appVersion: '1.2.14',
        createdAt: '2020-06-08T15:00:38.413+02:00',
        updatedAt: '2020-06-08T15:00:38.413+02:00',
    },
    {
        identifier: '1.0.12',
        json: '{"forms_version":"1.0.12","identifiers":["blood_screening.json"}',
        appId: 'org.smartregister.reveal',
        appVersion: '1.2.14',
        createdAt: '2020-06-08T15:03:00.950+02:00',
        updatedAt: '2020-06-08T15:03:00.950+02:00',
    },
    {
        identifier: '1.0.16',
        json: '{"forms_version":"1.0.16","identifiers":["blood_screening.json"}',
        appId: 'org.smartregister.reveal',
        appVersion: '1.2.16',
        createdAt: '2020-06-09T14:40:04.189+02:00',
        updatedAt: '2020-06-09T14:40:04.189+02:00',
    },
];
