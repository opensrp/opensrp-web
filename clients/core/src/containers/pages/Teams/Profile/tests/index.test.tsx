import { shallow, mount } from 'enzyme';
import { createLocation, createMemoryHistory } from 'history';
import * as React from 'react';
import { match, MemoryRouter } from 'react-router';
import ConnectedTeamProfile, { TeamProfile } from '../index';

import { Provider } from 'react-redux';
import store from '../../../../../store';
import teamMember, {
    removeTeamMembers,
    fetchTeamMembers,
    reducerName as teamMemberReducer,
} from '../../../../../store/ducks/teamMembers';
import * as fixtures from '../../../../../store/ducks/tests/fixtures';
import reducerRegistry from '@onaio/redux-reducer-registry';
import team, { fetchTeams, reducerName as teamsReducer } from '../../../../../store/ducks/teams';

// /** register the team reducer */
reducerRegistry.register(teamsReducer, team);

/** register the teamMember reducer */
reducerRegistry.register(teamMemberReducer, teamMember);

const history = createMemoryHistory();
const path = `/household-profile/:id/`;

const matchVariable: match<{ id: string }> = {
    isExact: false,
    params: { id: fixtures.team1.identifier },
    path,
    url: path.replace(':id', fixtures.team1.identifier),
};

const location = createLocation(matchVariable.url);

jest.mock('../../../../../configs/env');
describe('containers/households/Profile', () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let classMock: any;
    beforeEach(() => {
        jest.resetAllMocks();

        const memberListMock = jest.fn(async () => {
            return fixtures.teamMemberList;
        });

        const teamMock = jest.fn(async () => {
            return fixtures.team1;
        });

        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        classMock = jest.fn((a, b, c) => {
            return {
                list: b === 'organization' ? teamMock : memberListMock,
            };
        });
    });

    it('renders without crashing', () => {
        const props = {
            members: [],
            fetchTeam: fetchTeams,
            fetchMembers: fetchTeamMembers,
            removeMembers: removeTeamMembers,
            opensrpService: classMock,
            history,
            team: fixtures.team1,
            location,
            match: matchVariable,
        };
        const component = shallow(<TeamProfile {...props} />);
        const wrapper = component.find('#teamProfile');
        expect(wrapper.length).toBe(1);
    });

    it('should render the member table', () => {
        const props = {
            members: fixtures.teamMemberList,
            fetchTeam: fetchTeams,
            fetchMembers: fetchTeamMembers,
            removeMembers: removeTeamMembers,
            opensrpService: classMock,
            history,
            team: fixtures.team1,
            location,
            match: matchVariable,
        };
        const component = mount(
            <MemoryRouter initialEntries={['/']} keyLength={0}>
                <TeamProfile {...props} />
            </MemoryRouter>,
        );
        const wrapper = component.find('MemberTable');
        expect(wrapper.length).toBe(1);
    });

    it('should render when team is null', () => {
        const props = {
            members: [],
            fetchTeam: fetchTeams,
            fetchMembers: fetchTeamMembers,
            removeMembers: removeTeamMembers,
            opensrpService: classMock,
            history,
            team: null,
            location,
            match: matchVariable,
        };
        const component = shallow(<TeamProfile {...props} />);
        const wrapper = component.find('#teamProfile');
        expect(wrapper.length).toBe(0);
    });

    it('should work with store', async () => {
        store.dispatch(fetchTeams(fixtures.teamList));
        store.dispatch(fetchTeamMembers(fixtures.teamMemberList));
        const props = {
            members: [],
            fetchTeam: fetchTeams,
            fetchMembers: fetchTeamMembers,
            removeMembers: removeTeamMembers,
            opensrpService: classMock,
            history,
            team: null,
            location,
            match: matchVariable,
        };
        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter initialEntries={['/']} keyLength={0}>
                    <ConnectedTeamProfile {...props} />
                </MemoryRouter>
            </Provider>,
        );

        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        await new Promise((resolve: any) => setImmediate(resolve));
        wrapper.update();
        expect(wrapper).toMatchSnapshot();
        wrapper.unmount();
    });
});
