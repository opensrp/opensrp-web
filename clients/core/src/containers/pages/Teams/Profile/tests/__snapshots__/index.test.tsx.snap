// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/households/Profile should work with store 1`] = `ReactWrapper {}`;
